### ISSUE TITLE
[question] Checking if there is a conan command line in which it can package a library and all its dependencies into a single .tgz 

### ISSUE THREAD

**Issue by zxuan99** [conan-io/conan#17396]
### What is your question?

Hi! I have two questions.

I would like to check if there is a conan command line in which it can package a library and all its dependencies into a single .tgz to transfer it to my private server. I understand that there is a command "conan cache save "pkg/*:*""but it doesnt pack its dependencies into a single .tgz.

I have used conan install to install all the libraries into my internet facing computer and then transfer to my non-internet facing computer. However, in the visual studio IDE of my non-internet facing computer, it still complains about unable to retrieve the github package of the respective libraries. May I ask if there is a workaround for this issue? What command aside from conan install should I use to download all packages so that after I transfer these packages to my non-internet facing computer, it is still able to work?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @zxuan99 

> I would like to check if there is a conan command line in which it can package a library and all its dependencies into a single .tgz to transfer it to my private server. I understand that there is a command "conan cache save "pkg/:""but it doesnt pack its dependencies into a single .tgz.

The ``conan cache save`` stores all packages in the cache, but even older revisions, older versions, or packages that you are not using. the ``conan install --deployer`` will get the current used dependencies of your current project.

If you want to use the deployer to manage dependencies for current project ``--deployer=full_deploy`` you get a folder "full_deploy" with everything there. If you want to further automate that task, there are some possibilities:
- Create your own ``deployer``, that besides copying does an extra tgz.
- Create your own custom command that does the task

> I have used conan install to install all the libraries into my internet facing computer and then transfer to my non-internet facing computer. However, in the visual studio IDE of my non-internet facing computer, it still complains about unable to retrieve the github package of the respective libraries. May I ask if there is a workaround for this issue? What command aside from conan install should I use to download all packages so that after I transfer these packages to my non-internet facing computer, it is still able to work?

Have you considered building the packages instead from the ``conan-center-index`` Github repo. You can clone it and build from there, then upload the binaries to your own private server in the non-internet facing?

**Comment by zxuan99** [comment]
Thanks for your reply! In that case, combining the two issues that I faced in total, would you be able to advise what is the best way to download the packages and its dependencies that I need and then transfer it to my private server that is non-internet facing?

**Comment by memsharded** [comment]
> Thanks for your reply! In that case, combining the two issues that I faced in total, would you be able to advise what is the best way to download the packages and its dependencies that I need and then transfer it to my private server that is non-internet facing?

If you want to store dependencies in your private server, my recommendation would be to build them from sources from the ``conan-center-index`` repo, not to download the packages and upload them. Am I missing something?

Basically what is recommended here https://docs.conan.io/2/devops/conancenter/hosting_binaries.html

**Comment by zxuan99** [comment]
I see. I have missed that out. However, I am still unsure of the process since my private server is on a non internet facing computer. May I confirm the following from you?

I will first use an internet facing computer to "conan export" and "conan create --build=missing". Then I will transfer the binaries to the non internet facing computer in which my private server is and then upload it?

If that's the case, what is the ideal way to transfer the binaries of the libraries that I want and all its dependencies and upload it? Should I use the deployer as you mentioned just now to do it?

**Comment by memsharded** [comment]
> I will first use an internet facing computer to "conan export" and "conan create --build=missing". Then I will transfer the binaries to the non internet facing computer in which my private server is and then upload it?

You can build directly on the non-internet facing area. If we are talking about a corporate air-gapped system the recommendation is:

- Use ``conan-center-index`` repo
- For the dependencies you want, use the ``conandata.yml`` files to download the sources. Upload those sources to a generic repo in the server in the airgapped area. Modify the ``conandata.yml`` to point to your server. You can do this with a simple script
- Now do the ``conan export`` + ``conan create`` for the recipes you want, already in the air-gapped area. Sources will be downloaded from that area as well
- Upload to the server in the air-gap area

This is relatively simple to do, some short script.
This way you have:
- full control over the binaries, total security and ownership of the packages
- fully backuped sources so you can reproduce builds any point later in time
- Full control over the process of creating binaries, if you need to apply any customization to recipes, it is trivial as you already own the process, no need for ConanCenter to fix or merge something

**Comment by zxuan99** [comment]
Understood. If I have a library package and multiple package dependencies that needs to be uploaded into the server, is there a quick way to do it? Or is the only efficient way is to write a pkglist.json file myself and then use "conan upload --list=pkglist.json -r=myremote -c"

**Comment by memsharded** [comment]
> Understood. If I have a library package and multiple package dependencies that needs to be uploaded into the server, is there a quick way to do it? 

Package lists can be created from ``conan install`` too, like:
```
conan install ..... --format=json > graph.json
conan list -g graph.json --format=json > pkglist.json
```
You can even merge package lists with ``conan pkglist``

**Comment by zxuan99** [comment]
> > I will first use an internet facing computer to "conan export" and "conan create --build=missing". Then I will transfer the binaries to the non internet facing computer in which my private server is and then upload it?
> 
> You can build directly on the non-internet facing area. If we are talking about a corporate air-gapped system the recommendation is:
> 
> * Use `conan-center-index` repo
> * For the dependencies you want, use the `conandata.yml` files to download the sources. Upload those sources to a generic repo in the server in the airgapped area. Modify the `conandata.yml` to point to your server. You can do this with a simple script
> * Now do the `conan export` + `conan create` for the recipes you want, already in the air-gapped area. Sources will be downloaded from that area as well
> * Upload to the server in the air-gap area
> 
> This is relatively simple to do, some short script. This way you have:
> 
> * full control over the binaries, total security and ownership of the packages
> * fully backuped sources so you can reproduce builds any point later in time
> * Full control over the process of creating binaries, if you need to apply any customization to recipes, it is trivial as you already own the process, no need for ConanCenter to fix or merge something

Hi @memsharded,

I followed your recommendations. However, there is an issue that requires your help.

I am using Visual Studio with the conan server in an airgapped environment. However, when I tried to add dependency - rabbitmq-c/0.13.0 in the conandata.yml and build it, there is an error in Visual Studio - rabbitmq-c/0.13.01:(followed by a string of number and alphabets) folder must exist: C:\User\mycomputer\.conan2\p\b\rabbi97d8c7d381e64\p.

May I ask why it doesn't automatically fetch rabbitmq from the conan-server? What do I short of?

The error is only resolved when I used 'conan export' and 'conan create' for rabbitmq-c.

**Comment by memsharded** [comment]
I am not sure what you did. ``rabbitmq-c/0.13.0`` is not added to the conandata.yml. Only the sources from it are downloaded.
You would need to provide details:
- The files you are using
- The commands you are using
- The full output of the commands

**Comment by zxuan99** [comment]
Hi @memsharded,

I would like to ask you another question. I am confused about the phrase "binary packages". May I ask if nlohmann_json/3.11.2 is considered as a binary package? If so, I have upload it to the conan server in the air gapped environment using "conan upload" (according to https://docs.conan.io/2/reference/commands/upload.html - "By default, if a recipe reference is specified, it will upload all the revisions for all the binary packages") but then somehow it is not being uploaded to the server but it points to filepath in the condadata.yml.

**Comment by memsharded** [comment]
Hi @zxuan99 

The tutorial in https://docs.conan.io/2/tutorial/consuming_packages.html will cover these aspects.

Something like ``pkg/version`` can be used to reference to a "package", which is composed by 2 parts:

- A package **recipe**, with a reference like ``pkg/version#recipe_revision``
- One or multiple package binaries, with a reference like ``pkg/version#recipe_revision:package_id``. (package binaries also have revisions, but it is irrelevant for this discussion)


> By default, if a recipe reference is specified, it will upload all the revisions for all the binary packages") but then somehow it is not being uploaded to the server but it points to filepath in the condadata.yml.

I don't know what you mean, the ``conandata.yml`` is not involved in the ``conan upload`` process (besides being uploaded as any other file), so you would need to clarify what you mean, provide clear information with commands, the output of the commands, etc.

thanks for the feedback.

**Comment by zxuan99** [comment]
> > Thanks for your reply! In that case, combining the two issues that I faced in total, would you be able to advise what is the best way to download the packages and its dependencies that I need and then transfer it to my private server that is non-internet facing?
> 
> If you want to store dependencies in your private server, my recommendation would be to build them from sources from the `conan-center-index` repo, not to download the packages and upload them. Am I missing something?
> 
> Basically what is recommended here https://docs.conan.io/2/devops/conancenter/hosting_binaries.html

Hi @memsharded ,

I tried your recommendation which is to host the packages in my private server and acknowlege your point that it is better to build them from sources. However, I realize that there is an issue - several packages such as grpc and boost are taking extremely long time to build (more than 40 mins). For these packages, I intend to host them in binary instead of source so to cut down the build time. May I ask if you can advise on how to host the binary for these packages on the server?

**Comment by memsharded** [comment]
> I tried your recommendation which is to host the packages in my private server and acknowlege your point that it is better to build them from sources. However, I realize that there is an issue - several packages such as grpc and boost are taking extremely long time to build (more than 40 mins). For these packages, I intend to host them in binary instead of source so to cut down the build time. May I ask if you can advise on how to host the binary for these packages on the server?

Not only for these packages, but all packages are recommended to upload the binaries too, so it is not necessary to re-build from sources all the time.

The default ``conan upload`` behavior will already upload the binaries to the server, so typically is something as:

```
conan create .  # creates mypkg/myversion
conan upload mypkg/myversion -r=myremote -c
```

So when someone does ``conan install`` it will also by default download the pre-compiled binary from the server and it is not necessary to build it from source. Doing this is the default and recommended approach.

**Comment by memsharded** [comment]
Any further question or comment here?
Thanks for the feedback!

**Comment by memsharded** [comment]
Closing this ticket as responded (and inactive).
Don't hesitate to comment or to create new tickets for any further question or issue you might have. Thanks for your feedback.

**Comment by patrick-fromberg** [comment]
> This is relatively simple to do, some short script.

Or so you say regarding downloading from the urls in conandata.yml. Simple to do, sure, but impossibly hard to maintain I would say. If you purge conan packages you have to replicate that on your download store as well. Its like conan needs another package manager for raw downloads. Conan users need to implement this helper package manager themselves. 

It should not be to hard to modify all conanfile.py files such that the downloaded data is stored in the sources directory of the package. I did exactly that with a single script modifying all conanfiles in my conan cache and this works for most packages but not for all, sadly. But modifying recipes pior to uploading them internally is not pretty and trying to reverse engineer conan to get something that is obviously needed and not provided by conan is frustrating. The conan team could provide this out of the box without conanfile modifications needed. End users can not do this. I am not happy myself with build systems fulfilling every special kink and consequently becoming over complicated. But packages containing full sources instead of referencing that information in the internet is something that I am certain all corporate users need. Auditor: "Where are your sources stored? In your Package Manager?", User: "No, they are somewhere in the internet, but I cannot recall all the places where"

BTW, my first experiment was enforcing every one to enable the download cache. My modified conanfile.py would copy the content from the download cache into the package if not found there. The newly created package would do the opposite. So the cache data was stored in the packages but ultimately consumed from the download cache that was now fed from the package sources. That worked fairly well but would probably not support cross platform builds. The reason why I mentions this is, that without this experiment I would not have noticed that not all downloaded data goes into the conan cache. You learn conan by making errors, not by reading the documentation. 

As of now, I see no viable solution for air-gaped systems. The best I can do is to have all external dependencies prebuilt on an internet facing machine and simply anticipate all possible conan profiles in use, which In my case are only four. This solves the building in air-gaped  environment problem but not the auditing problems.

**Comment by memsharded** [comment]
Hi @patrick-fromberg 

Thanks for the feedback.

I think there are different things in your comment, and as this is a closed ticket with a title that doesn't match what you are describing that well, it would be probably better to open new tickets. Maybe more than one, with more focused/targeted feedback or feature requests.

For example:

> As of now, I see no viable solution for air-gaped systems. The best I can do is to have all external dependencies prebuilt on an internet facing machine and simply anticipate all possible conan profiles in use, which In my case are only four. This solves the building in air-gaped environment problem but not the auditing problems.

I am not sure if I understand this. We are already aware of quite a few users using Conan in completely air-gapped systems.
For example using a ``conan graph info -c tools.build:download_source=True`` + ``conan cache backup-upload`` seems to be working quite well, and as the sources are the same (at least for ConanCenter packages) in the vast majority of cases, using only 1 profile is good. Isn't this the case for your use case?

> BTW, my first experiment was enforcing every one to enable the download cache. My modified conanfile.py would copy the content from the download cache into the package if not found there. The newly created package would do the opposite. So the cache data was stored in the packages but ultimately consumed from the download cache that was now fed from the package sources.

I don't understand this either, copying the sources into the final package in the ``package()`` method? Then not clear how the sources are used from there.


> Or so you say regarding downloading from the urls in conandata.yml. Simple to do, sure, but impossibly hard to maintain I would say. If you purge conan packages you have to replicate that on your download store as well. Its like conan needs another package manager for raw downloads. Conan users need to implement this helper package manager themselves.

Not clear where is the maintenance problem here. We have also seen this as a popular way of implementing this before the backup-sources feature was implemented, and talked about it with many users before implementing the feature, maintenance didn't seem a problem at all, like another script that cross-check the cached sources against the recipes repo.


In summary, I think it would be better to try to understand better the possible open pain points, and for that it would be better to do it in new tickets, that focus on one aspect, for example if we are talking about the Conan backup-sources features, try to understand why the download + conan cache backup-upload

**Comment by patrick-fromberg** [comment]
I am not opening a new issue here as you suggest because there is already such an issue, but it is also closed and I dont have permission to reopen it. I think the issue of air-gaping is far from being closed and not only because of my comments and questions. When I say the issue is open, I dont mean the documentation only.  

I did not know about `conan graph info -c tools.build:download_source=True + conan cache backup-upload`. This looks useful indeed but the documentation is lacking on the details.

- The documentation says: "Additionally, these sources can be transparently uploaded alongside the packages via conan upload.". Does this mean that if I build a package and upload it, then the cache will also be uploaded implicitly?
- The documentation says that if I do not set a source cache location, then the default source cache location is used. But my observation is that I can either set a sources download cache location or downloads will not be cached at all. There is no default source download location as far as I can see. BTW, I am using conan 2.23.0.
- Regarding `conan cache backup-upload`. Totally unclear what this does in the context of the next questions.
- According to the documentation `get()` and `download()` calls will be cached in the source download cache. That implies that `git.fetch_commit()` calls are not. But experimentation shows me that `git.fetch_commit()` is also cached, but elsewhere. I delete the sources download cache via `rm ~/.conan2/<path_of_my_source_cache>` and I clean the cache for the git downloads via `conan cache clean */*:*`. So which cache is covered by `conan cache backup-upload`? Why are there two different caches in the first place? There are even two more caches, but my understanding is they have both been deprecated.
- 
The only working solution as far as I understand:
- git repository for cci fork.
- nexus or artifactory conan server for the packages
- nexus or artifactory server (raw data) for downloaded sources
- nexus or artifactory server (raw data) for metadata because metadata is not official and consequently not supported by Nexus (what about Artifactory?).

Because this works it is also the solution that probably most air-gaped projects use. That is not the same as being a "popular" solution. Those who do conan for a living are blind to the fact that this is very cumbersome.  

I still believe the cleanest solution is to allow users to add the downloaded sources to the package in a convenient way. Example: `conan download --bundle-sources ...` or even better, a flag `bundle-sources=True` in the `global.config`. The hashing algorithm for the package can be tweaked to ignore these added sources which have already their own sha256. 

I can see why conancenter only wants to store references to sources but there is no target conflict if users and conancenter want different things here. What conancenter stores and what finally lands in the packages on the users side after downloading does not need to be the same thing (only the hashes must be the same). 

I bet users that dont need an air-gaped solution would surely be happy with air-gaped packages as well because they are conceptually simpler. No need for bothering users with different options here. If different packages referencing the same sources are a problem, then there are better ways to deal with those, i.e. ways that store duplicated data only once and hide that fact.

What I tried myself may not make sense but it is good enough as a proof of concept I believe. Here is what I injected into every `conanfile.py` at the start of the `source()` method in the case the download uses `git.fetch_commit()` ( I had something similar for get() ):
```
        # BEGIN INJECTED
        _base = _os.path.dirname(self.source_folder) 
        _bundled = _os.path.join(_base, '_git_source')
        if _os.path.exists(_bundled):
            for _item in _os.listdir(_bundled):
                _src = _os.path.join(_bundled, _item)
                _dst = _os.path.join(self.source_folder, _item)
                if _os.path.isdir(_src):
                    _shutil.copytree(_src, _dst)
                else:
                    _shutil.copy2(_src, _dst)
            return
        # END INJECTED
        git = Git(self)
        git.fetch_commit(**self.conan_data["sources"][self.version])
        apply_conandata_patches(self)
```
Note that the script injecting this code also copied the source code from the cache to the package. All this worked for most packages but not for boost for e.g. and some others. I am not a conan specialist and as a c++ dev I should not be required to be one, especially not a specialist for conan internals.

**Comment by memsharded** [comment]
> I am not opening a new issue here as you suggest because there is already such an issue, but it is also closed and I dont have permission to reopen it. I think the issue of air-gaping is far from being closed and not only because of my comments and questions. When I say the issue is open, I dont mean the documentation only.

Then, please open a completely new ticket, keeping discussions on closed tickets is simply more complicated and more likely to be dropped.

> nexus or artifactory server (raw data) for metadata because metadata is not official and consequently not supported by Nexus (what about Artifactory?).

Not sure what you mean here. Do you mean Conan metadata files? Like https://blog.conan.io/2023/10/24/Conan-launches-metadata-files.html? those should work with any server that implements the Conan protocol, because it doesn't use any special server side api or anything like that, files are managed within the existing Conan protocol.


> Because this works it is also the solution that probably most air-gaped projects use. That is not the same as being a "popular" solution. Those who do conan for a living are blind to the fact that this is very cumbersome.

It doesn't work that way. We design features around the feedback we get from users. We even do calls very frequently with different users to directly learn from their needs, how they use Conan ,what they are missing, etc, and we design, implement and deliver new features based on that feedback. It is "popular", because I know from direct feedback that there are quite a few of those users we talked with about it that they are happy about it, because it is what they wanted. Maybe it is cumbersome to some other users, but those users working on those air-gapped systems (btw, from different industries, different sizes, etc.) are happy about it.

That doesn't mean that there is no space for further improvements, and we are happy to keep learning, discovering and considering new flows, tools and functionalities to keep improving.
But it would be way better to do that in a new open ticket that doesn't diverge this much from the main issue of the ticket, as this is a closed ticket on a question about bundling dependencies, while this discussion is about doing a copy of the sources in recipes

**Comment by patrick-fromberg** [comment]
Thanks for responding during holidays. If this is what users want, then I regard this issue as closed as well. I am very opposed to serve users with very special requests, it complicates matters for everyone in the end. I was not aware of me being "special" in this regard.
